File: python/src/main/java/org/locationtech/geowave/python/Debug.java
/**
 * Copyright (c) 2013-2022 Contributors to the Eclipse Foundation
 *
 * <p> See the NOTICE file distributed with this work for additional information regarding copyright
 * ownership. All rights reserved. This program and the accompanying materials are made available
 * under the terms of the Apache License, Version 2.0 which accompanies this distribution and is
 * available at http://www.apache.org/licenses/LICENSE-2.0.txt
 */
package org.locationtech.geowave.python;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

/**
 * A class for debugging tools.
 */
public class Debug {

  /**
   * Prints information about the object on both python (returns a String) and java side.
   * 
   * @param obj
   */
  public String printObject(Object obj) {
    System.out.println(obj);
    return obj.toString();
  }

  /**
   * Prints (verbose) information about the object on both python (returns a String) and java side.
   * 
   * @param obj
   * @param verbose
   */
  public String printObject(Object obj, boolean verbose) {
    if (!verbose) {
      return printObject(obj);
    }

    StringBuilder methods = new StringBuilder();

    for (Method method : obj.getClass().getMethods()) {
      methods.append(method.getName()).append(" ;");
    }

    StringBuilder fields = new StringBuilder();

    for (Field field : obj.getClass().getFields()) {
      fields.append(field.getName()).append("; ");
    }
    StringBuilder info = new StringBuilder();
    info.append("Object: ").append(obj.toString()).append("\n").append("Class: ").append(
        obj.getClass().toString()).append("\n").append("isNull: ").append(obj == null).append(
            "\n").append("Methods: ").append(methods.toString()).append("\n").append(
                "Fields: ").append(fields.toString()).append("\n");

    System.out.println(info.toString());
    return info.toString();
  }
}

File: python/src/main/python/pygw/debug.py
#
# Copyright (c) 2013-2020 Contributors to the Eclipse Foundation

#
# See the NOTICE file distributed with this work for additional information regarding copyright
# ownership. All rights reserved. This program and the accompanying materials are made available
# under the terms of the Apache License, Version 2.0 which accompanies this distribution and is
# available at http://www.apache.org/licenses/LICENSE-2.0.txt
#===============================================================================================
"""
This module exposes a function called `print_obj` that can be used to help with debugging raw java objects. It will print information about the object in question on both the Python side and on the Java server side. There's a `verbose` flag that will give you more information about the object in question.
"""
from pygw.config import java_gateway
from pygw.base import GeoWaveObject

def print_obj(to_print, verbose=False):
    """
    Print method to help with debugging.
    """
    if isinstance(to_print, GeoWaveObject):
        to_print = to_print._java_ref
    print(java_gateway.entry_point.getDebug().printObject(to_print, verbose))
